src/device/device_ledger.cpp

```diff
@@ -317,6 +317,30 @@ namespace hw {
       memset(this->buffer_recv, 0, BUFFER_RECV_SIZE);
     }
 
+    int device_ledger::prologue(BYTE type, BYTE extra) {
+      reset_buffer();
+      int offset = 0;
+      this->buffer_send[0] = 0x00;
+      this->buffer_send[1] = type;
+      this->buffer_send[2] = extra;
+      this->buffer_send[3] = 0x00;
+      this->buffer_send[4] = 0x00;
+      return 5;
+    }
+
+    int device_ledger::prologue_noopt(BYTE type, BYTE extra) {
```

Comment:
Why is it necessary to have `prologue` and `prologue_noopt` separately? Can't we just have them like:
```c++
int device_ledger::prologue(BYTE type, BYTE extra) {
  reset_buffer();
  this->buffer_send[0] = 0x00;
  this->buffer_send[1] = type;
  this->buffer_send[2] = extra;
  this->buffer_send[3] = 0x00;
  this->buffer_send[4] = 0x00;
  int offset = 5;
  //options
  this->buffer_send[offset++] = 0;
  return offset;
}

void device_ledger::send_simple(BYTE type, BYTE extra) {
  int offset = prologue(type, extra);
  this->buffer_send[4] = offset - 5;
  this->length_send = offset;
  this->exchange();
}
```

Reply:
Just to allow some more factoring while avoiding a function with (what I think would be) too many parameters.

Reply:
But this way of split seems weird to me, because IIUI this line
```c++
this->buffer_send[4] = offset - 5;
```
is supposed to be paired with these lines
```c++
      this->length_send = offset;
      this->exchange();
```
I.e. the content of `buffer_send[4]` indicates the size of optional argument data starting from `buffer_send[5]`. In the current patch, the call to `prologue_noopt()` puts 1 to `buffer_send[4]`, which will then be overwritten by the subsequent line `this->buffer_send[4] = offset-5;`. The end result is the same, but this patch adds an unnecessary step of `buffer_send[4] = 1` which seems weird to me. I also don't really understand what you said above (to allow more factoring and to avoiding a function with too many parameters). Can you elaborate?

Reply:
OK, thanks for the clarification.